### ISSUE TITLE
pqsql protocol sending "immense" fields

### ISSUE THREAD

**Issue by portante** [elastic/beats#195]
Have one instance where I am periodically seeing the following in the ES logs for a system where we have Postgres running.

```
[2015-08-11 19:42:56,514][DEBUG][action.bulk              ] [gronk-gprfs004] [packetbeat-2015.08.11][0] failed to execute bulk item (index) index {[packetbeat-2015.08.11][pgsql][AU8eSXu5n0Xwxxv0Ri30], source[{"bytes_in":140125,"bytes_out":486,"client_ip":"10.12.23.226","client_port":40532,"client_proc":"","client_server":"","count":1,"ip":"10.12.23.114","method":"SELECT","pgsql":{"error_code":"","error_message":"","error_severity":"","iserror":false,"num_fields":17,"num_rows":0},"port":5432,"proc":"","query":"SELECT \"snapshots\".* FROM \"snapshots\"  WHERE \"snapshots\".\"vm_or_template_id\" IN (80000, 79999, 79998, 79997, 79996, 79995, 79994, 79993, 79992, 79991, ...
java.lang.IllegalArgumentException: Document contains at least one immense term in field="query" (whose UTF8 encoding is longer than the max length 32766), all of which were skipped.  Please correct the analyzer to not produce such terms.  The prefix of the first immense term is: '[83, 69, 76, 69, 67, 84, 32, 34, 115, 110, 97, 112, 115, 104, 111, 116, 115, 34, 46, 42, 32, 70, 82, 79, 77, 32, 34, 115, 110, 97]...', original message: bytes can be at most 32766 in length; got 140119
        at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:687)  
        at org.apache.lucene.index.DefaultIndexingChain.processField(DefaultIndexingChain.java:359)  
        at org.apache.lucene.index.DefaultIndexingChain.processDocument(DefaultIndexingChain.java:318)  
        at org.apache.lucene.index.DocumentsWriterPerThread.updateDocument(DocumentsWriterPerThread.java:241)  
        at org.apache.lucene.index.DocumentsWriter.updateDocument(DocumentsWriter.java:465)  
        at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1526)  
        at org.apache.lucene.index.IndexWriter.addDocument(IndexWriter.java:1252)  
        at org.elasticsearch.index.engine.InternalEngine.innerCreateNoLock(InternalEngine.java:343)  
        at org.elasticsearch.index.engine.InternalEngine.innerCreate(InternalEngine.java:285)  
        at org.elasticsearch.index.engine.InternalEngine.create(InternalEngine.java:256)  
        at org.elasticsearch.index.shard.IndexShard.create(IndexShard.java:470)  
        at org.elasticsearch.action.bulk.TransportShardBulkAction.shardIndexOperation(TransportShardBulkAction.java:437)  
        at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnPrimary(TransportShardBulkAction.java:149)  
        at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$AsyncShardOperationAction.performOnPrimary(TransportShardReplicationOperationAction.java:515)  
        at org.elasticsearch.action.support.replication.TransportShardReplicationOperationAction$AsyncShardOperationAction$1.run(TransportShardReplicationOperationAction.java:422)  
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)  
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)  
        at java.lang.Thread.run(Thread.java:745)  
Caused by: org.apache.lucene.util.BytesRefHash$MaxBytesLengthExceededException: bytes can be at most 32766 in length; got 140119  
        at org.apache.lucene.util.BytesRefHash.add(BytesRefHash.java:284)  
        at org.apache.lucene.index.TermsHashPerField.add(TermsHashPerField.java:151)  
        at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:663)  
        ... 17 more
```

**Comment by tsg** [comment]
Thanks for the report, we need to guard against such huge fields in Packetbeat. In this case, the `query` field is huge because it contains probably a huge list of numbers. Because `query` is mapped as not analyzed in Elasticsearch, the whole thing is considered a single term, and ES rejects terms over 32K.

**Comment by portante** [comment]
@tsg, maybe we need to write a simple "analyzer" for sql query strings to pull out the basic parts, yet allow long queries to be indexed.

**Comment by tsg** [comment]
You mean an Elasticsearch analyzer? Or as a analyzer in Packetbeat/libbeat?

I think it would be nice for the queries to be parsed and simplified by Packetbeat for other reasons as well, like grouping similar queries together automatically. Not sure if there's an easy way to do that reliably, but it can resort to trimming if things don't seem to work. The original query can still be found in the `request` field if it's enabled. The `request` field is mapped as analyzed, so it doesn't have this issue.

**Comment by portante** [comment]
@tsg, yes, an ElasticSearch analyzer field.  The BNF of a SQL query is pretty well defined, so it should be fairly straightforward to write an analyzer for it.

**Comment by kbrock** [comment]
actually parsing SQL is a little tricky, but using a few regular expressions may simplify the sql and be good enough.

Rails framework does simplify the queries when writing to the logs. Wonder if you could leverage that work, either by using their string as input and not writing a parser, or seeing their approach and leveraging some of the work.

**Comment by tsg** [comment]
@portante I was wondering if when this happened you had the [template](https://github.com/elastic/packetbeat/blob/master/etc/packetbeat.template.json) loaded. We set `ingore_above: 1024`, so actually I think the term shouldn't have been indexed.

**Comment by portante** [comment]
I'll look ...

**Comment by portante** [comment]
I have had the template loaded since I installed packetbeat.

I have not seen it since September 22nd, but that might be because the system that was using postgres was changed or reinstalled.  Not sure.

It appears that I did not update to 1.0.0-beta4 until October 12th or so.

I'll be updating to -rc1 today for all our systems, dropping our existing indexes, reinstalling the template, and then we can see if this happens again.
